vault/external_tests/plugin/plugin_test.go

```diff
@@ -0,0 +1,1073 @@
+package plugin_test
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/builtin/plugin"
+	"github.com/hashicorp/vault/helper/namespace"
+	vaulthttp "github.com/hashicorp/vault/http"
+	"github.com/hashicorp/vault/sdk/helper/consts"
+	"github.com/hashicorp/vault/sdk/helper/pluginutil"
+	"github.com/hashicorp/vault/sdk/logical"
+	lplugin "github.com/hashicorp/vault/sdk/plugin"
+	"github.com/hashicorp/vault/sdk/plugin/mock"
+	"github.com/hashicorp/vault/vault"
+)
+
+const (
+	expectedEnvKey   = "FOO"
+	expectedEnvValue = "BAR"
+)
+
+// logicalVersionMap is a map of version to test plugin
+var logicalVersionMap = map[string]string{
+	"v4":             "TestBackend_PluginMain_V4_Logical",
+	"v5":             "TestBackend_PluginMainLogical",
+	"v5_multiplexed": "TestBackend_PluginMain_Multiplexed_Logical",
+}
+
+// credentialVersionMap is a map of version to test plugin
+var credentialVersionMap = map[string]string{
+	"v4":             "TestBackend_PluginMain_V4_Credentials",
+	"v5":             "TestBackend_PluginMainCredentials",
+	"v5_multiplexed": "TestBackend_PluginMain_Multiplexed_Credentials",
+}
+
+var testCtx = context.TODO()
+
+func TestSystemBackend_Plugin_secret(t *testing.T) {
+	t.Parallel()
+	testCases := []struct {
+		pluginVersion string
+	}{
+		{
+			pluginVersion: "v5_multiplexed",
+		},
+		{
+			pluginVersion: "v5",
+		},
+		{
+			pluginVersion: "v4",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.pluginVersion, func(t *testing.T) {
+			t.Parallel()
+			cluster := testSystemBackendMock(t, 1, 1, logical.TypeLogical, tc.pluginVersion)
+			defer cluster.Cleanup()
+
+			core := cluster.Cores[0]
+
+			// Make a request to lazy load the plugin
+			req := logical.TestRequest(t, logical.ReadOperation, "mock-0/internal")
+			req.ClientToken = core.Client.Token()
+			resp, err := core.HandleRequest(namespace.RootContext(testCtx), req)
+			if err != nil {
+				t.Fatalf("err: %v", err)
+			}
+			if resp == nil {
+				t.Fatalf("bad: response should not be nil")
+			}
+
+			// Seal the cluster
+			cluster.EnsureCoresSealed(t)
+
+			// Unseal the cluster
+			barrierKeys := cluster.BarrierKeys
```

Comment:
I am wondering why not use `cluster.EnsureCoresUnsealed` here?

Reply:
I'm just moving funcs around, I didn't write these tests.